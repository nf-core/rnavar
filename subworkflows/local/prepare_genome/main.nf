//
// Prepare reference genome files
//

include { BEDTOOLS_MERGE                    } from '../../../modules/nf-core/bedtools/merge/main'
include { BEDTOOLS_SORT                     } from '../../../modules/nf-core/bedtools/sort/main'
include { GATK4_CREATESEQUENCEDICTIONARY    } from '../../../modules/nf-core/gatk4/createsequencedictionary/main'
include { GFFREAD                           } from '../../../modules/nf-core/gffread/main'
include { GTF2BED                           } from '../../../modules/local/gtf2bed'
include { SAMTOOLS_FAIDX                    } from '../../../modules/nf-core/samtools/faidx/main'
include { STAR_GENOMEGENERATE               } from '../../../modules/nf-core/star/genomegenerate/main'
include { GUNZIP as GUNZIP_FASTA            } from '../../../modules/nf-core/gunzip/main'
include { GUNZIP as GUNZIP_GTF              } from '../../../modules/nf-core/gunzip/main'
include { TABIX_TABIX as TABIX_DBSNP        } from '../../../modules/nf-core/tabix/tabix/main'
include { TABIX_TABIX as TABIX_KNOWN_INDELS } from '../../../modules/nf-core/tabix/tabix/main'

workflow PREPARE_GENOME {
    take:
    fasta_raw      // file: /path/to/genome.fasta
    gff            // file: /path/to/genome.gff
    gtf_raw        // file: /path/to/genome.gtf
    dbsnp
    known_indels
    feature_type

    main:
    ch_versions = Channel.empty()

    //Unzip reference genome files if needed

    if (params.fasta.endsWith('.gz')) {
        GUNZIP_FASTA(fasta_raw)

        fasta = GUNZIP_FASTA.out.gunzip

    } else {
        fasta = fasta_raw
    }

    if (params.gtf.endsWith('.gz')) {
        GUNZIP_GTF(gtf_raw)

        gtf = GUNZIP_GTF.out.gunzip

    } else {
        gtf = gtf_raw
    }

    GATK4_CREATESEQUENCEDICTIONARY(fasta)
    GFFREAD(gff, fasta)
    SAMTOOLS_FAIDX(fasta, [['id':'genome'], []])
    TABIX_DBSNP(dbsnp.flatten().map{ it -> [ [ id:it.baseName ], it ] })
    TABIX_KNOWN_INDELS(known_indels.flatten().map{ it -> [ [ id:it.baseName ], it ] } )

    gtf = gtf.mix(GFFREAD.out.gtf)

    GTF2BED(gtf, feature_type)
    STAR_GENOMEGENERATE(fasta, gtf)

    ch_versions = ch_versions.mix(GATK4_CREATESEQUENCEDICTIONARY.out.versions)
    ch_versions = ch_versions.mix(GFFREAD.out.versions)
    ch_versions = ch_versions.mix(GTF2BED.out.versions)
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)
    ch_versions = ch_versions.mix(STAR_GENOMEGENERATE.out.versions)
    ch_versions = ch_versions.mix(TABIX_DBSNP.out.versions)
    ch_versions = ch_versions.mix(TABIX_KNOWN_INDELS.out.versions)

    emit:
    dict             = GATK4_CREATESEQUENCEDICTIONARY.out.dict                              //    path: genome.fasta.dict
    exon_bed         = GTF2BED.out.bed.map{ bed -> [ [ id:bed.baseName ], bed ] }.collect() //    path: exon.bed
    fasta            = fasta
    fasta_fai        = SAMTOOLS_FAIDX.out.fai                                               //    path: genome.fasta.fai
    gtf              = gtf.first()                                                       //    path: genome.gtf
    star_index       = STAR_GENOMEGENERATE.out.index.first()                                //    path: star/index/
    dbsnp_tbi        = TABIX_DBSNP.out.tbi.map{ _meta, tbi -> [tbi] }.collect()              // path: dbsnb.vcf.gz.tbi
    known_indels_tbi = TABIX_KNOWN_INDELS.out.tbi.map{ _meta, tbi -> [tbi] }.collect()       // path: {known_indels*}.vcf.gz.tbi
    versions         = ch_versions                                                          // channel: [ versions.yml ]
    // bedtools_sort    = bedtools_sort    // path: sort.bed
    // bedtools_merge   = bedtools_merge   // path: merge.bed
}
