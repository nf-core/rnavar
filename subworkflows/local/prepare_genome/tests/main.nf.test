nextflow_workflow {

    name "Test Subworkflow PREPARE_GENOME"
    script "../main.nf"
    workflow "PREPARE_GENOME"

    tag "subworkflows"
    tag "subworkflows_local"

    test("only essentials") {
        when {
            workflow {
                """
                input[0] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                input[1] = []
                input[2] = []
                input[3] = []
                input[4] = []
                input[5] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true)
                input[6] = []
                input[7] = []
                input[8] = []
                input[9] = []
                input[10] = []
                input[11] = []
                input[12] = []
                input[13] = "exon"
                input[14] = false
                input[15] = Channel.value(false)
                """
            }
        }

        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        } else if(key == "star_index") {
                            return [key, val.collect { [it[0], path(it[1]).list().collect { file(it.toString()).name }] }]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }

    test("all") {
        when {
            workflow {
                """
                input[0] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                input[1] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                input[2] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                input[3] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/index/star/star.tar.gz', checkIfExists: true)
                input[4] = []
                input[5] = file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true)
                input[6] = file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.bed', checkIfExists: true)
                input[7] = file(params.modules_testdata_base_path + '/genomics/sarscov2/illumina/vcf/test2.vcf.gz', checkIfExists: true)
                input[8] = []
                input[9] = file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf', checkIfExists: true)
                input[10] = []
                input[11] = file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz', checkIfExists: true)
                input[12] = file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi', checkIfExists: true)
                input[13] = "exon"
                input[14] = false
                input[15] = Channel.value(true)
                """
            }
        }

        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        } else if(key == "star_index") {
                            return [key, val.collect { [it[0], path(it[1]).list().collect { file(it.toString()).name }] }]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }
}
