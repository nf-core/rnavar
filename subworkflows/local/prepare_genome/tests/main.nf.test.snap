{
    "all": {
        "content": [
            {
                "dbsnp": [
                    [
                        {
                            "id": "dbsnp"
                        },
                        "dbsnp.vcf.gz:md5,1bb8724dcbe6fa3101a814c0be51d1ea"
                    ]
                ],
                "dbsnp_tbi": [
                    [
                        {
                            "id": "dbsnp"
                        },
                        "dbsnp.vcf.gz.tbi:md5,747c6529930c04120f367110a9ea5e63"
                    ]
                ],
                "dict": [
                    [
                        {
                            "id": "dict"
                        },
                        "/nf-core/test-datasets/rnavar/data/genomics/homo_sapiens/genome/genome.dict"
                    ]
                ],
                "exon_bed": [
                    [
                        {
                            "id": "exon_bed"
                        },
                        "exon_bed.exon.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "fasta"
                        },
                        "/nf-core/test-datasets/rnavar/data/genomics/homo_sapiens/genome/genome.fasta"
                    ]
                ],
                "fasta_fai": [
                    [
                        {
                            "id": "fasta"
                        },
                        "genome.fasta.fai:md5,3520cd30e1b100e55f578db9c855f685"
                    ]
                ],
                "gtf": [
                    [
                        {
                            "id": "gtf"
                        },
                        "/nf-core/test-datasets/rnavar/data/genomics/homo_sapiens/genome/genome.gtf"
                    ]
                ],
                "known_indels": [
                    [
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz"
                    ]
                ],
                "known_indels_tbi": [
                    [
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi"
                    ]
                ],
                "star_index": [
                    [
                        {
                            "id": "star"
                        },
                        [
                            "Genome",
                            "Log.out",
                            "SA",
                            "SAindex",
                            "chrLength.txt",
                            "chrName.txt",
                            "chrNameLength.txt",
                            "chrStart.txt",
                            "exonGeTrInfo.tab",
                            "exonInfo.tab",
                            "geneInfo.tab",
                            "genomeParameters.txt",
                            "sjdbInfo.txt",
                            "sjdbList.fromGTF.out.tab",
                            "sjdbList.out.tab",
                            "transcriptInfo.tab"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2f8adf49812faab29b90f617e24e7ddc",
                    "versions.yml:md5,3c6fc57c4c3bf7b143933b0b89ab0386",
                    "versions.yml:md5,7263c03ca9c7d18bc40fc8d27a5936fa",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12",
                    "versions.yml:md5,f104f9bec6db3d55c728ffd041131110"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-30T14:37:56.543826207"
    },
    "only essentials": {
        "content": [
            {
                "dbsnp": [
                    [
                        [
                            
                        ],
                        [
                            
                        ]
                    ]
                ],
                "dbsnp_tbi": [
                    [
                        [
                            
                        ],
                        [
                            
                        ]
                    ]
                ],
                "dict": [
                    [
                        {
                            "id": "fasta"
                        },
                        "genome.dict:md5,2433fe2ba31257337bf4c4bd4cb8da15"
                    ]
                ],
                "exon_bed": [
                    [
                        {
                            "id": "gtf"
                        },
                        "gtf.exon.bed:md5,17192418c24c524881e7d970f94a4a26"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "fasta"
                        },
                        "/nf-core/test-datasets/rnavar/data/genomics/homo_sapiens/genome/genome.fasta"
                    ]
                ],
                "fasta_fai": [
                    [
                        {
                            "id": "fasta"
                        },
                        "genome.fasta.fai:md5,3520cd30e1b100e55f578db9c855f685"
                    ]
                ],
                "gtf": [
                    [
                        {
                            "id": "gtf"
                        },
                        "/nf-core/test-datasets/rnavar/data/genomics/homo_sapiens/genome/genome.gtf"
                    ]
                ],
                "known_indels": [
                    
                ],
                "known_indels_tbi": [
                    
                ],
                "star_index": [
                    
                ],
                "versions": [
                    "versions.yml:md5,265795dcfabc74e982198a30abb773d5",
                    "versions.yml:md5,2f8adf49812faab29b90f617e24e7ddc",
                    "versions.yml:md5,3c6fc57c4c3bf7b143933b0b89ab0386",
                    "versions.yml:md5,5a1daa8526f6fedf8e0e9a2453184119",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-30T14:37:41.563216435"
    }
}